src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -270,7 +264,10 @@ Klass* SystemDictionary::resolve_or_fail(Symbol* class_name, Handle class_loader
                                          bool throw_error, TRAPS) {
   Klass* klass = resolve_or_null(class_name, class_loader, protection_domain, THREAD);
   // Check for pending exception or null klass, and throw exception
-  return handle_resolution_exception(class_name, throw_error, klass, THREAD);
+  if (HAS_PENDING_EXCEPTION || klass == NULL) {
+    handle_resolution_exception(class_name, throw_error, CHECK_NULL);
+  }
```

Comment:
I find the logic for this very confusing. If klass == NULL then isn't it the case that there must be a pending exception?
Also, looking inside handle_resolution_exception, if the pending exception is not a ClassNotFoundException then we completely ignore it and throw either ClassNotFoundException or NoClassDefFoundError. I would have expected to at least chain the original exception.

Reply:
No, a klass can be NULL if load_instance_class returns NULL in a user-defined class loader.
ClassLoader.loadClass() can return NULL.
    // Class is not found or has the wrong name, return NULL
    return NULL;
There's also a NULL return from the bootclasspath.
For instance when loading sun/launcher/resources/spi/launcherProvider, this package isn't yet in the bootloader's list of packages, so there's a NULL return from load_instance_class, which is returned by resolve_instance_class_or_null.

It is sort of confusing that resolve_instance_class_or_null can return an exception OR null.  if it returns an exception, the exception and throw_error is true, the exception is wrapped in the NoClassDefFoundError.  It's in the JVMS 5.3:

If the Java Virtual Machine ever attempts to load a class C during verification (§5.4.1) or resolution (§5.4.3) (but not initialization (§5.5)), and the class loader that is used to initiate loading of C throws an instance of ClassNotFoundException, then the Java Virtual Machine must throw an instance of NoClassDefFoundError whose cause is the instance of ClassNotFoundException.

We don't ignore the exception if it's not ClassNotFoundException, we don't clear the exception so it's returned by handle_resolution_exception.

Reply:
`// Class is not found or has the wrong name, return NULL`

If Class is not found then there should be a ClassNotFoundException, or NoClassDefFoundError pending.
If the representation has the wrong name then again a NoClassDefFoundError should be pending.

However, I see there are some paths in `load_instance_class` related to the module system that will return NULL without there being a pending exception.

Reply:
Yes, it would be clearer if load_instance_class returned an exception and not NULL.  I'll note this as a further cleanup because it is inconsistent.  There are many calls to SystemDictionary::resolve_or_null that expect NULL and not a pending exception though so there are a lot of paths through this code that all want different answers. I don't know how to improve it but I'll look at it.